include/gatekeeper_ggu.h

```diff
@@ -47,35 +47,77 @@ struct ggu_config {
 	struct mailbox    mailbox;
 };
 
+/* Enumeration of policy decisions the GT-GK block can process. */
+enum {
+	/* Decline an IPv4 flow. */
+	GGU_DEC_IPV4_DECLINED = 0,
+	/* Decline an IPv6 flow. */
+	GGU_DEC_IPV6_DECLINED,
+	/* Grant an IPv4 flow. */
+	GGU_DEC_IPV4_GRANTED,
+	/* Grant an IPv6 flow. */
+	GGU_DEC_IPV6_GRANTED,
+	__MAX_GGU_DEC
+};
+
+/* On the wire policy decision to be processed by the GT-GK Unit. */
+struct ggu_decision {
+	/* The policy decision type. */
+	uint16_t type;
+	/* The length of the policy decision, including any parameters. */
+	uint16_t len;
+
+	/* The IP flow relevant to this policy decision. */
+	uint8_t  ip_flow[0];
+
+	/* Parameters for this policy decision would follow the IP flow. */
+}__attribute__((packed));
+
 /*
- * Since the length of IPv6 policy is much larger than IPv4 policy,
- * to save network bandwidth, we choose to use different data structures
- * to store the in-packet policies for IPv4 and IPv6, respectively.
+ * Packets that flow between the GT block and the GT-GK Unit.
+ * 
+ * Packet format: Ethernet header(s) + IP header + UDP header + Data.
+ * The UDP payload data is of the following format:
  *
- * Packet format: Ethernet headers + IP header + UDP header + Data.
- * In the UDP payload, the following format would save a lot of bytes
- * when there are decline decisions:
- *  v1, n1, n2, n3, n4: Each of these fields are 1-byte long.
- *  v1 is a constant indicating the version of the format, in this case 1.
- *  n1 is the number of IPv4 decline decisions.
- *  n2 is the number of IPv6 decline decisions.
- *  n3 is the number of IPv4 granted decisions.
- *  n4 is the number of IPv6 granted decisions.
+ *  v1: a constant indicating the version of the format, in this case 1
+ *  reserved: reserved space to keep the fields 32-bit aligned
+ *  [policy decisions]: a list of one or more policy decisions
  * 
- * Field v1 will enable us to change the format, incrementally update
- * the Gatekeeper servers, and incrementally update the Grantor servers.
+ * Each policy decision (struct ggu_decision) is of the following format:
+ *  
+ * +---------------------------------+
+ * |  Decision Type  |    Length     |
+ * +---------------------------------+
+ * |                                 |
+ * |  IP flow (source, destination)  |
+ * |                ...              |
+ * +---------------------------------+
+ * |                                 |
+ * |        Decision Parameters      |
+ * |                ...              |
+ * +---------------------------------+
+ *
+ * The decision type is from an enumerated set that both the GT block
+ * and the GT-GK Unit must understand in order for the decision to be
+ * processed. For example, an action may grant an IPv4 flow. Policy
+ * decisions that the GT-GK Unit does not understand will be skipped.
+ *
+ * The length of the decision is for validating that the decision contains
```

Comment:
I think we don't need the length field now. As the decision can be validated by the checksum of the headers. Moreover, if there is a policy decision that cannot be understood by the GGU, shouldn't we drop that packet instead of skipping it?